pkg/ds/farm.go

```diff
@@ -0,0 +1,411 @@
+package ds
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/square/p2/pkg/util"
+
+	"github.com/Sirupsen/logrus"
+	"github.com/square/p2/pkg/alerting"
+	"github.com/square/p2/pkg/ds/fields"
+	ds_fields "github.com/square/p2/pkg/ds/fields"
+	"github.com/square/p2/pkg/kp"
+	"github.com/square/p2/pkg/kp/dsstore"
+	"github.com/square/p2/pkg/labels"
+	"github.com/square/p2/pkg/logging"
+	"github.com/square/p2/pkg/scheduler"
+	"github.com/square/p2/pkg/types"
+	klabels "k8s.io/kubernetes/pkg/labels"
+)
+
+// Farm instatiates and deletes daemon sets as needed
+type Farm struct {
+	// constructor arguments
+	kpStore    kp.Store
+	dsStore    dsstore.Store
+	scheduler  scheduler.Scheduler
+	applicator labels.Applicator
+
+	children map[fields.ID]*childDS
+	childMu  sync.Mutex
+
+	logger  logging.Logger
+	alerter alerting.Alerter
+}
+
+type childDS struct {
+	ds        DaemonSet
+	quitCh    chan<- struct{}
+	updatedCh chan<- *ds_fields.DaemonSet
+	deletedCh chan<- *ds_fields.DaemonSet
+	errCh     <-chan error
+}
+
+func NewFarm(
+	kpStore kp.Store,
+	dsStore dsstore.Store,
+	applicator labels.Applicator,
+	logger logging.Logger,
+	alerter alerting.Alerter,
+) *Farm {
+	if alerter == nil {
+		alerter = alerting.NewNop()
+	}
+
+	return &Farm{
+		kpStore:    kpStore,
+		dsStore:    dsStore,
+		scheduler:  scheduler.NewApplicatorScheduler(applicator),
+		applicator: applicator,
+		children:   make(map[fields.ID]*childDS),
+		logger:     logger,
+		alerter:    alerter,
+	}
+}
+
+func (dsf *Farm) Start(quitCh <-chan struct{}) {
+	dsf.cleanupDaemonSetPods(quitCh)
+	dsf.mainLoop(quitCh)
+}
+
+const cleanupInterval = 60 * time.Second
+
+// This function removes all pods with a DSIDLabel where the daemon set id does
+// not exist in the store at every interval specified because it is possible
+// that the farm will unexpectedly crash or someone deletes or modifies a node
+func (dsf *Farm) cleanupDaemonSetPods(quitCh <-chan struct{}) {
+	go func() {
+		timer := time.NewTimer(time.Duration(0))
+
+		for {
+			select {
+			case <-quitCh:
+				return
+			case <-timer.C:
+			}
+			timer.Reset(cleanupInterval)
+
+			allDaemonSets, err := dsf.dsStore.List()
+			if err != nil {
+				dsf.logger.NoFields().Errorf("Unable to get daemon sets from intent tree: %v", err)
```

Comment:
I suspect we're going to want some fields on all log messages, such as daemon set ID, pod ID, availability zone, etc. Otherwise it will be hard to determine the impact of these errors